libgadget/tests/test_gravity.c

```diff
@@ -0,0 +1,364 @@
+{
+
+    double r2 = 0;
+    int d;
+    double dist[3];
+    for(d = 0; d < 3; d ++) {
+        /* the distance vector points to 'other' */
+        dist[d] = offset[d] + P[this].Pos[d] - P[other].Pos[d];
+        r2 += dist[d] * dist[d];
+    }
+
+    const double r = sqrt(r2);
+
+    const double h = FORCE_SOFTENING(1);
+
+    double fac = 1 / (r2 * r);
+    if(r < h) {
+        double h_inv = 1.0 / h;
+        double h3_inv = h_inv * h_inv * h_inv;
+        double u = r * h_inv;
+        if(u < 0.5)
+            fac = 1. * h3_inv * (10.666666666667 + u * u * (32.0 * u - 38.4));
+        else
+            fac =
+                1. * h3_inv * (21.333333333333 - 48.0 * u +
+                        38.4 * u * u - 10.666666666667 * u * u * u - 0.066666666667 / (u * u * u));
+    }
+
+    for(d = 0; d < 3; d ++) {
+        accns[3*this + d] += - dist[d] * fac * All.G * P[other].Mass;
+        accns[3*other + d] += dist[d] * fac * All.G * P[this].Mass;
+    }
+}
+
+void check_accns(double * meanerr_tot, double * maxerr_tot, double *PairAccn, double meanacc)
+{
+    double meanerr=0, maxerr=-1;
+    int i;
+    /* This checks that the short-range force accuracy is being correctly estimated.*/
+    #pragma omp parallel for reduction(+: meanerr) reduction(max: maxerr)
+    for(i = 0; i < PartManager->NumPart; i++)
+    {
+        int k;
+        for(k=0; k<3; k++) {
+            double err = fabs((PairAccn[3*i+k] - (P[i].GravPM[k] + P[i].GravAccel[k]))/meanacc);
+            meanerr += err;
+            if(maxerr < err)
+                maxerr = err;
+        }
+    }
+    MPI_Allreduce(&meanerr, meanerr_tot, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
+    MPI_Allreduce(&maxerr, maxerr_tot, 1, MPI_DOUBLE, MPI_MAX, MPI_COMM_WORLD);
+    int64_t tot_npart;
+    sumup_large_ints(1, &PartManager->NumPart, &tot_npart);
+
+    *meanerr_tot/= (tot_npart*3.);
+}
+
+static void find_means(double * meangrav, double * suppmean, double * suppaccns)
+{
+    int i;
+    double meanacc = 0, meanforce = 0;
+    #pragma omp parallel for reduction(+: meanacc) reduction(+: meanforce)
+    for(i = 0; i < PartManager->NumPart; i++)
+    {
+        int k;
+        for(k=0; k<3; k++) {
+            if(suppaccns)
+                meanacc += fabs(suppaccns[3*i+k]);
+            meanforce += fabs(P[i].GravPM[k] + P[i].GravAccel[k]);
+        }
+    }
+    int64_t tot_npart;
+    sumup_large_ints(1, &PartManager->NumPart, &tot_npart);
+    if(suppaccns) {
+        MPI_Allreduce(&meanacc, suppmean, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
+        *suppmean/= (tot_npart*3.);
+    }
+    MPI_Allreduce(&meanforce, meangrav, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
+    *meangrav/= (tot_npart*3.);
+}
+
+/* This checks the force on each particle using a direct summation:
+ * very slow, but accurate.
+ * Periodic boundary conditions are included by mirroring the box.*/
+static int force_direct(double ErrTolForceAcc)
```

Comment:
There might be some difficulty computing force directly after particles are scattered -- but you can always gather them to the same rank and compute the force.

Do we have a routine that can sort particles in P by ID or alike? If there is a stable ordering, then you can literally compute the residual between direct force and TreePM and report the variance etc?

Reply:
What do you mean by that? Isn't that what I do in check_accns?

Reply:
I see. check_accns computes the residual. 

Two questions remains:
1. how does the direct_force calculation work across ranks? Looks like the loop does not include particles from other ranks.
2. is is possible to rename force_direct to check_force_againast_direct, and lift the direct force calculation loop to a function force_direct?

Reply:
1. It does not - the test is serial only at the moment. This should be fixed, along with the density test (and perhaps the forcetree test), and I'll put that on my todo list. For this one I would indeed probably gather them to one rank.

2. Yes, I will do that!